File: src/go/radio/type_def_view.go
package radio

// ViewMaterialChangeCommand — 修改 mesh 可视化材质的指令（完整版）
type ViewMaterialChangeCommand struct {
	MaterialType              *string     `json:"materialType,omitempty"`
	Color                     *string     `json:"color,omitempty"`
	Opacity                   *float64    `json:"opacity,omitempty"`
	Transparent               *bool       `json:"transparent,omitempty"`
	Visible                   *bool       `json:"visible,omitempty"`
	Wireframe                 *bool       `json:"wireframe,omitempty"`
	Side                      *int        `json:"side,omitempty"`
	DepthTest                 *bool       `json:"depthTest,omitempty"`
	DepthWrite                *bool       `json:"depthWrite,omitempty"`
	AlphaTest                 *float64    `json:"alphaTest,omitempty"`
	Blending                  *int        `json:"blending,omitempty"`
	VertexColors              *bool       `json:"vertexColors,omitempty"`
	Fog                       *bool       `json:"fog,omitempty"`
	Emissive                  *string     `json:"emissive,omitempty"`
	EmissiveIntensity         *float64    `json:"emissiveIntensity,omitempty"`
	Metalness                 *float64    `json:"metalness,omitempty"`
	Roughness                 *float64    `json:"roughness,omitempty"`
	EnvMapIntensity           *float64    `json:"envMapIntensity,omitempty"`
	FlatShading               *bool       `json:"flatShading,omitempty"`
	Shininess                 *float64    `json:"shininess,omitempty"`
	Specular                  *string     `json:"specular,omitempty"`
	Clearcoat                 *float64    `json:"clearcoat,omitempty"`
	ClearcoatRoughness        *float64    `json:"clearcoatRoughness,omitempty"`
	Transmission              *float64    `json:"transmission,omitempty"`
	Ior                       *float64    `json:"ior,omitempty"`
	Thickness                 *float64    `json:"thickness,omitempty"`
	Sheen                     *float64    `json:"sheen,omitempty"`
	SheenRoughness            *float64    `json:"sheenRoughness,omitempty"`
	SheenColor                *string     `json:"sheenColor,omitempty"`
	AttenuationColor          *string     `json:"attenuationColor,omitempty"`
	AttenuationDistance       *float64    `json:"attenuationDistance,omitempty"`
	Iridescence               *float64    `json:"iridescence,omitempty"`
	IridescenceIOR            *float64    `json:"iridescenceIOR,omitempty"`
	IridescenceThicknessRange *[2]float64 `json:"iridescenceThicknessRange,omitempty"`
	SpecularIntensity         *float64    `json:"specularIntensity,omitempty"`
	SpecularColor             *string     `json:"specularColor,omitempty"`
	Reflectivity              *float64    `json:"reflectivity,omitempty"`
	Dispersion                *float64    `json:"dispersion,omitempty"`
	Anisotropy                *float64    `json:"anisotropy,omitempty"`
	AnisotropyRotation        *float64    `json:"anisotropyRotation,omitempty"`
	MapURL                    *string     `json:"mapUrl,omitempty"`
	NormalMapURL              *string     `json:"normalMapUrl,omitempty"`
	RoughnessMapURL           *string     `json:"roughnessMapUrl,omitempty"`
	MetalnessMapURL           *string     `json:"metalnessMapUrl,omitempty"`
	EmissiveMapURL            *string     `json:"emissiveMapUrl,omitempty"`
	AoMapURL                  *string     `json:"aoMapUrl,omitempty"`
	AlphaMapURL               *string     `json:"alphaMapUrl,omitempty"`
	BumpMapURL                *string     `json:"bumpMapUrl,omitempty"`
	DisplacementMapURL        *string     `json:"displacementMapUrl,omitempty"`
	NormalScale               *[2]float64 `json:"normalScale,omitempty"`
	BumpScale                 *float64    `json:"bumpScale,omitempty"`
	DisplacementScale         *float64    `json:"displacementScale,omitempty"`
	DisplacementBias          *float64    `json:"displacementBias,omitempty"`
	AoMapIntensity            *float64    `json:"aoMapIntensity,omitempty"`
	MapRepeat                 *[2]float64 `json:"mapRepeat,omitempty"`
	MapOffset                 *[2]float64 `json:"mapOffset,omitempty"`
	MapRotation               *float64    `json:"mapRotation,omitempty"`
}

// ViewMaterialChangeCommandSimple — 修改 mesh 可视化材质的指令（简化版）
type ViewMaterialChangeCommandSimple struct {
	Color             *string  `json:"color,omitempty"`
	Opacity           *float64 `json:"opacity,omitempty"`
	Transparent       *bool    `json:"transparent,omitempty"`
	Visible           *bool    `json:"visible,omitempty"`
	Wireframe         *bool    `json:"wireframe,omitempty"`
	Side              *int     `json:"side,omitempty"`
	Fog               *bool    `json:"fog,omitempty"`
	Emissive          *string  `json:"emissive,omitempty"`
	EmissiveIntensity *float64 `json:"emissiveIntensity,omitempty"`
}

File: src/rs/radio/type_def_view.rs
use serde::{Deserialize, Serialize};

/// 修改 mesh 的可视化材质的指令（完整版）
#[derive(Debug, Clone, Serialize, Deserialize, Default)]
#[serde(rename_all = "camelCase")]
pub struct ViewMaterialChangeCommand {
    #[serde(skip_serializing_if = "Option::is_none")]
    pub material_type: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub color: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub opacity: Option<f64>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub transparent: Option<bool>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub visible: Option<bool>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub wireframe: Option<bool>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub side: Option<i32>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub depth_test: Option<bool>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub depth_write: Option<bool>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub alpha_test: Option<f64>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub blending: Option<i32>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub vertex_colors: Option<bool>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub fog: Option<bool>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub emissive: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub emissive_intensity: Option<f64>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub metalness: Option<f64>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub roughness: Option<f64>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub env_map_intensity: Option<f64>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub flat_shading: Option<bool>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub shininess: Option<f64>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub specular: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub clearcoat: Option<f64>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub clearcoat_roughness: Option<f64>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub transmission: Option<f64>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub ior: Option<f64>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub thickness: Option<f64>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub sheen: Option<f64>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub sheen_roughness: Option<f64>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub sheen_color: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub attenuation_color: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub attenuation_distance: Option<f64>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub iridescence: Option<f64>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub iridescence_ior: Option<f64>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub iridescence_thickness_range: Option<(f64, f64)>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub specular_intensity: Option<f64>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub specular_color: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub reflectivity: Option<f64>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub dispersion: Option<f64>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub anisotropy: Option<f64>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub anisotropy_rotation: Option<f64>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub map_url: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub normal_map_url: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub roughness_map_url: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub metalness_map_url: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub emissive_map_url: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub ao_map_url: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub alpha_map_url: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub bump_map_url: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub displacement_map_url: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub normal_scale: Option<(f64, f64)>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub bump_scale: Option<f64>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub displacement_scale: Option<f64>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub displacement_bias: Option<f64>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub ao_map_intensity: Option<f64>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub map_repeat: Option<(f64, f64)>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub map_offset: Option<(f64, f64)>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub map_rotation: Option<f64>,
}

/// 修改 mesh 的可视化材质的指令（简化版）
#[derive(Debug, Clone, Serialize, Deserialize, Default)]
#[serde(rename_all = "camelCase")]
pub struct ViewMaterialChangeCommandSimple {
    #[serde(skip_serializing_if = "Option::is_none")]
    pub color: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub opacity: Option<f64>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub transparent: Option<bool>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub visible: Option<bool>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub wireframe: Option<bool>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub side: Option<i32>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub fog: Option<bool>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub emissive: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub emissive_intensity: Option<f64>,
}
